File: src/test/java/com/github/jactor/web/model/UserModelTest.java
package com.github.jactor.web.model;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;

import com.github.jactor.web.dto.AddressDtoBuilder;
import com.github.jactor.web.dto.PersonDtoBuilder;
import com.github.jactor.web.dto.UserDtoBuilder;
import java.util.List;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

@DisplayName("A UserModel")
class UserModelTest {

  @Test
  @DisplayName("should fetch the address of the user as a list of strings")
  void shouldFetchTheAddress() {
    UserModel testUserModel = new UserModel(
        aUser().with(aPerson().with(anAddress()
            .withAddressLine1("address line 1")
            .withAddressLine2("address line 2")
            .withAddressLine3("address line 3")
            .withZipCode("1234")
            .withCity("somewhere")
        )).build()
    );

    List<String> address = testUserModel.fetchAddress();

    assertThat(address).containsExactly(
        "address line 1",
        "address line 2",
        "address line 3",
        "1234",
        "somewhere"
    );
  }

  @Test
  @DisplayName("should not fetch parts of address being null")
  void shouldGetTheAddress() {
    UserModel testUserModel = new UserModel(
        aUser().with(aPerson().with(anAddress()
            .withAddressLine1("address line")
            .withZipCode("4321")
        )).build()
    );

    List<String> address = testUserModel.fetchAddress();

    assertThat(address).containsExactly(
        "address line",
        "4321"
    );
  }

  @Test
  @DisplayName("should fetch the username of the user")
  void shouldFetchTheUsername() {
    UserModel testUserModel = new UserModel(aUser().withUsername("user").build());

    assertThat(testUserModel.fetchUsername()).isEqualTo("user");
  }

  @Test
  @DisplayName("should fetch the person behind the user")
  void shouldFetchThePersonBehindTheUser() {
    UserModel testUserModel = new UserModel(
        aUser().with(aPerson()
            .withFirstName("John")
            .withSurname("Smith")
            .withDescription("description")
        ).build()
    );

    assertAll(
        () -> assertThat(testUserModel.fetchFullName()).isEqualTo("John Smith"),
        () -> assertThat(testUserModel.fetchDescriptionCode()).isEqualTo("description")
    );
  }

  private UserDtoBuilder aUser() {
    return new UserDtoBuilder();
  }

  private PersonDtoBuilder aPerson() {
    return new PersonDtoBuilder();
  }

  private AddressDtoBuilder anAddress() {
    return new AddressDtoBuilder();
  }

}

File: src/test/kotlin/com/github/jactor/web/model/UserModelTest.kt
package com.github.jactor.web.model

import com.github.jactor.shared.dto.AddressDto
import com.github.jactor.shared.dto.PersonDto
import com.github.jactor.shared.dto.UserDto
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertAll

internal class UserModelTest {
    @Test
    fun `should fetch the address of the user as a list of strings`() {
        val testUserModel = UserModel(
            UserDto(
                person = PersonDto(
                    address = AddressDto(
                        addressLine1 = "address line 1",
                        addressLine2 = "address line 2",
                        addressLine3 = "address line 3",
                        zipCode = "1234",
                        city = "somewhere"
                    )
                )
            )
        )

        val address = testUserModel.fetchAddress()
        assertThat(address).containsExactly(
            "address line 1",
            "address line 2",
            "address line 3",
            "1234",
            "somewhere"
        )
    }

    @Test
    fun `should not fetch parts of address being null`() {
        val testUserModel = UserModel(UserDto(person = PersonDto(address = AddressDto(addressLine1 = "address line 1", zipCode = "1234"))))

        val address = testUserModel.fetchAddress()
        assertThat(address).containsExactly(
            "address line 1",
            "1234"
        )
    }

    @Test
    fun `should fetch the username of the user`() {
        val testUserModel = UserModel(UserDto(username = "user"))
        assertThat(testUserModel.fetchUsername()).isEqualTo("user")
    }

    @Test
    fun `should fetch the person behind the user`() {
        val testUserModel = UserModel(UserDto(person = PersonDto(firstName = "John", surname = "Smith", description = "description")))

        assertAll(
            { assertThat(testUserModel.fetchFullName()).isEqualTo("John Smith") },
            { assertThat(testUserModel.fetchDescriptionCode()).isEqualTo("description") }
        )
    }
}
